internal/evmreader/claim.go

```diff
@@ -0,0 +1,224 @@
+				"last claim check block", lastClaimCheck,
+				"most recent block", mostRecentBlockNumber,
+			)
+		}
+
+	}
+}
+
+func (r *EvmReader) readAndUpdateClaims(
+	ctx context.Context,
+	apps []application,
+	lastClaimCheck, mostRecentBlockNumber uint64,
+) {
+
+	// DISCLAIMER: The current algorithm will only handle Authority.
+	// To handle Quorum, node needs to handle acceptance events
+	// that can happen before claim submission
+
+	// Classify them by same IConsensusAddress
+	sameConsensusApps := indexApps(keyByIConsensus, apps)
+	for iConsensusAddress, apps := range sameConsensusApps {
+
+		appAddresses := appToAddresses(apps)
+
+		// All these apps shares the same IConsensus
+		if len(apps) == 0 {
+			continue
+		}
+		consensusContract := apps[0].consensusContract
+
+		// Retrieve Claim Acceptance Events from blockchain
+		appClaimAcceptanceEventMap, err := r.readClaimAcceptanceFromBlockchain(
+			ctx, consensusContract, appAddresses, lastClaimCheck+1, mostRecentBlockNumber)
+		if err != nil {
+			slog.Error("Error reading claim acceptance status",
+				"apps", apps,
+				"IConsensus", iConsensusAddress,
+				"start", lastClaimCheck,
+				"end", mostRecentBlockNumber,
+				"error", err)
+			continue
+		}
+
+		// Check events against Epochs
+	APP_LOOP:
+		for app, claimAcceptances := range appClaimAcceptanceEventMap {
+
+			epochs := []*Epoch{}
+			for _, claimAcceptance := range claimAcceptances {
+
+				// Get Previous Epochs with submitted claims, that got no acceptance event
+				// and update their statuses rejecting them all
+				previousEpochs, err := r.repository.GetPreviousSubmittedClaimsEpochs(
+					ctx, app, claimAcceptance.LastProcessedBlockNumber.Uint64())
+				if err != nil {
+					slog.Error("Error retrieving previous submitted claims",
+						"app", app, "error", err)
+					continue APP_LOOP
+				}
+
+				for _, previousEpoch := range previousEpochs {
+					previousEpoch.Status = EpochStatusClaimRejected
+					epochs = append(epochs, previousEpoch)
+					slog.Warn("Claim rejected",
+						"app", app,
+						"lastBlock", previousEpoch.LastBlock,
+						"hash", previousEpoch.ClaimHash)
+				}
+
+				// Get the Epoch for the current Claim Acceptance Event
+				epoch, err := r.repository.GetEpoch(
+					ctx, calculateEpochIndex(
+						r.epochLengthCache[app],
+						claimAcceptance.LastProcessedBlockNumber.Uint64()),
+					app)
+				if err != nil {
+					slog.Error("Error retrieving Epoch", "app", app, "error", err)
+					continue APP_LOOP
+				}
+
+				// Check Epoch
+				if epoch == nil {
+					slog.Error("Got claim acceptance event for an unknown epoch",
+						"app", app,
+						"claim last block", claimAcceptance.LastProcessedBlockNumber,
+						"hash", claimAcceptance.Claim)
+					continue APP_LOOP
+				}
+
+				// Update Epoch claim status
+				if claimAcceptance.Claim != *epoch.ClaimHash ||
+					claimAcceptance.LastProcessedBlockNumber.Uint64() != epoch.LastBlock {
+					slog.Warn("Claim Rejected",
```

Comment:
Another unexpected scenario that should be signaled properly
```suggestion
					slog.Warn("Claim Mismatch",
```